RFC-0003-functions-stats.md

```diff
@@ -0,0 +1,373 @@
+/// provide additional information regarding how this function impacts following
+/// query statistics.
+struct VectorFunctionStatistics {
+  /// Disable or enable propagate stats behaviour. A function (e.g.
+  /// upper(Slice)) may not alter any source stats for the input column passed
+  /// as an argument. In this case it might be possible to just set propagate
+  /// stats as true. In cases where number of input columns are more than one,
+  /// one has few choices i.e. is the resulting statistics a union of all
+  /// columns, Max, Sum or an intersection. To achieve this behaviour please set
+  /// `statsResolver` field in conjunction of setting this. For more precise
+  /// control consider supplying a callback via ... see Phase 3.
+  bool propagateStats{false};
+
+  /// Possible values: Max, Sum, union and intersect.
+  std::string statsResolver{"Max"};
+
+  /// Distinct values count is another important statistic in measuring query
+  /// perf characteristics, Does this function produces a constant Distinct
+  /// value count regardless of `input column`'s source stats.
+  double distinctValuesCount{NAN};
+
+  /// distinctValCountAdjustFactor: a fraction multiplied to input column's
+  /// distinctValuesCount.
+  double distinctValCountAdjustFactor{1.0};
+
+  /// Does this function produce a constant nullFraction, e.g. is_null(Slice)
+  /// will alter column's null fraction value to 0.0.
+  double nullFraction{NAN};
+
+  /// A `nullFraction` Fraction of column's entries that are null and
+  /// nullFractionAdjustFactor is the number that is multiplied to input
+  /// column's nullFraction value to compute this functions impact on overall
+  /// `nullFraction`.
+  double nullFractionAdjustFactor{1.0};
+
+  /// An `avgRowSize`: does this function impacts the size of each row e.g. a
+  /// function like md5 may produce a constant row size.
+  double avgRowSize{NAN};
+
+  /// An `avgRowSizeAdjustFactor`: does this function impacts the size of each
+  /// row e.g. a function may alter the column type to a wider type
+  /// e.g.varchar(20)->varchar(40).
+  double avgRowSizeAdjustFactor{1.0};
+};
+```
+
+Add a virtual function to `VectorFunctions.h` which a function can override with the function specific 
+statistics propagation override, e.g. 
+
+```cpp
+@@ -109,6 +160,11 @@ class VectorFunction {
+     return false;
+   }
+ 
++  // How this function impacts statistics.
++  virtual std::optional<VectorFunctionStatistics> statsResolver() const {
++    return std::nullopt;
++  }
++
+```
+
+```cpp
+@@ -129,6 +129,10 @@ class UpperLowerTemplateFunction : public exec::VectorFunction {
+   bool propagateStringEncodingFromAllInputs() const override {
+     return true;
+   }
++
++  std::optional<exec::VectorFunctionStatistics> statsResolver() const override {
++    return std::optional{exec::VectorFunctionStatistics{true}};
++  }
+ };
+```
+
+ * JAVA: Provide a new Annotation i.e. `ScalarFunctionStatsCalculator` in java with fields as follows:
```

Comment:
In CPP it looks like you're providing a function. This would potentially allow for more complex implementations in the future such as the sampling approach described earlier. With just an annotation on the scalar function class on the Java side, I don't believe that would be possible.

Do you think it would be possible to annotate a java function within the scalar function definition to allow custom logic other than `propagate=true`? Or provide a custom calculator implementation?

Reply:
Hi @ZacBlanco , thanks for the feedback and I am still working on it.

In CPP even though it seems to be a function, but it is not a very powerful abstraction. The abstraction does not allow a way to capture the stats of the input columns and transform them. It is slightly better than JAVA's annotation.

> Do you think it would be possible to annotate a java function within the scalar function definition to allow custom logic other than propagate=true? Or provide a custom calculator implementation?

Yes, I have a plan for it i.e. Phase 3 of implementation.